File: 1101-1199/1176/1176.c
#include <stdio.h>

int n, m, k = 0;

void dfs(int i, int a[][n], int b[])
{
	int j;
	for (j=0; j<n; j++)
		if (i != j && a[i][j] == 0)
			{ a[i][j] = 1; dfs(j, a, b); }
	b[k++] = i;
}

int main()
{
	int i, j;
	scanf("%d %d", &n, &m);
	
	int a[n][n];
	int b[n*n];

	for (i=0; i<n; i++)
		for (j=0; j<n; j++)
			scanf("%d", &a[i][j]);
	
	dfs(m-1, a, b);
	
	for (i=k-1; i>0; i--)
		printf("%d %d\n", b[i]+1, b[i-1]+1);
			
	return 0;
}

File: 1101-1199/1176/1176.py
import sys

n, m = [int(x) for x in sys.stdin.readline().split()]
a = []

for i in range(n):
	a.append([int(x) for x in sys.stdin.readline().split()])

ans = []

def dfs(i):
	for j in range(n):
		if i != j and a[i][j] == 0:
			a[i][j] = 1
			dfs(j)
	ans.append(i)

def dfs2(i):
	stack = [(i,0)]
	while len(stack) > 0:
		i, j = stack[-1]
		while j < n:
			if i != j and a[i][j] == 0:
				a[i][j] = 1
				stack.append((j,0))
				break
			else:
				j += 1
		if j == n:
			stack.pop()
			ans.append(i)

dfs2(m-1)

for i in range(len(ans)-1, 0, -1):
	print ans[i]+1, ans[i-1]+1
